### DISCUSSION TITLE
Unable to use a custom converter

### DISCUSSION THREAD

**Discussion by lloydjatkinson** [AvaloniaUI/Avalonia#15073] [Q&A]
I am unable to use a converter in XAML. I followed [the documentation][1] however it seems outdated.

```csharp
namespace Mazes.Ui.Temporary.Converters
{
    public class TimeSpanConverter : IValueConverter
    {
        public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
        {
           // ...
        }
    }
}
```

I register the converter in `MainWindow` where the window is declared.

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:Mazes.Ui.Temporary.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:views="clr-namespace:Mazes.Ui.Temporary.Views"
        xmlns:local="clr-namespace:Mazes.Ui.Temporary.Converters;assembly=Mazes.Ui.Temporary"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="Mazes.Ui.Temporary.Views.MainWindow"
        Icon="/Assets/avalonia-logo.ico"
        Title="Mazes.Ui.Temporary">
    <Window.Resources>
        <local:TimeSpanConverter x:Key="timeSpanConverter" />
    </Window.Resources>
    <views:MainView />
</Window>
```

I then try to use it in `MainView` where it is bound to a property in the view model.

```xaml
<TextBlock Text="{CompiledBinding ComputeTime, Converter={StaticResource timeSpanConverter}}" Grid.Row="0" Grid.Column="1"/>
```

However I get the following error in both `MainView` and `MainWindow`: `Static resource 'timespanConverter' not found.`

How should I register this custom converter? I have a simple reproduction of this problem here, that contains only an example converter: https://github.com/lloydjatkinson/throwaway

  [1]: https://docs.avaloniaui.net/docs/guides/data-binding/how-to-create-a-custom-data-binding-converter

**Comment by maxkatz6** [comment]
Documentation is not outdated, but you have found a corner case with static resources.
StaticResource is resolved when control is created, in your case - MainView. When MainView is created, it's not yet attached to the parent yet, thus - there is no timeSpanConverter visible from the MainView.

You either need to add timeSpanConverter to the MainView (where it's used) or to the Application Resources (which are globally visible).

**Reply by maxkatz6** [reply]
Or you can use something like x:Static, which I personally prefer:
https://docs.avaloniaui.net/docs/guides/data-binding/how-to-create-a-custom-data-binding-converter#example-1

**Reply by lloydjatkinson** [reply]
Thank you! I had understood the docs to mean that declaring it once for a Window would make it avaliable for everything else. How would I specifically add it to application resources? What's the difference between that and doing it the `x:Static` way?

**Reply by timunie** [reply]
Example for x:Static https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/MVVM/ValueConversionSample#solution-3--using-funcvalueconverter
